localstack-core/localstack/aws/protocol/serializer.py

```diff
@@ -614,6 +657,30 @@ def _add_md5_header(self, response: Response):
     def _get_error_message(self, error: Exception) -> str | None:
         return str(error) if error is not None and str(error) != "None" else None
 
+    def _get_error_status_code(
+        self, error: ServiceException, headers: Headers, service_model: ServiceModel
+    ) -> int:
+        # by default, some protocols (namely `json` and `smithy-rpc-v2-cbor`) might not define exception status code in
+        # their specs, so they are not defined in the `ServiceException` object and will use the default value of `400`
+        # But Query compatible service always do define them, so we get the wrong code for service that are
+        # multi-protocols like CloudWatch
+        # we need to verify if the service is compatible, and if the client has requested the query compatible error
+        # code to return the right value
+        if not service_model.is_query_compatible:
+            return error.status_code
+
+        if self._is_request_query_compatible(headers):
+            return error.status_code
+
+        # we only want to override status code 4XX
+        if 400 < error.status_code <= 499:
+            return 400
+
+        return error.status_code
+
+    def _is_request_query_compatible(self, headers: Headers) -> bool:
```

Comment:
nit: will update the type hint of headers as above, and will update the if check to be `headers and headers.get(...`, commit is ready, waiting for the full run